### ISSUE TITLE
Sik-Ho Tang | Review: AlexNet, CaffeNet -- Winner of ILSVRC 2012 (Image Classification).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#83]
Sik-Ho Tang. [Review: AlexNet, CaffeNet — Winner of ILSVRC 2012 (Image Classification)](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160).

**Comment by NorbertZheng** [comment]
## Overview
In this story, AlexNet and CaffeNet are reviewed. AlexNet is the winner of the ILSVRC ([ImageNet Large Scale Visual Recognition Competition](http://www.image-net.org/challenges/LSVRC/)) 2012, which is an image classification competition.

This is a 2012 NIPS paper from Prof. Hinton’s Group with about 28000 citations when I was writing this story. It has an essential breakthrough in deep learning which substantially reduce the error rate in ILSVRC 2012 as the figure shown below. Thus, this is a must read paper!!

 - ImageNet, is a dataset of over 15 millions labeled high-resolution images with around 22,000 categories.

ILSVRC uses a subset of ImageNet of around 1000 images in each of 1000 categories. In all, there are roughly 1.2 million training images, 50,000 validation images and 150,000 testing images.

![image](https://user-images.githubusercontent.com/37235023/236661374-0fb9a7c6-c42e-42f9-b4a0-836e5340ad49.png)
AlexNet, the winner in ILSVRC 2012 image classification with remarkable lower error rate.

A. For AlexNet, we will cover:
 - Architecture
 - ReLU (Rectified Linear Unit)
 - Multiple GPUs
 - Local Response Normalization
 - Overlapping Pooling
 - Data Augmentation
 - Dropout
 - Other Details of Learning Parameters
 - Results

B. For CaffeNet, it is just a single-GPU version of AlexNet. Since normally, people would only have one GPU, CaffeNet is a single-GPU network to simulate AlexNet. We will cover this as well at the end of this story.

By going through each component, we can know the importance of each component. Some of them are not so useful by now. But they do inspire for invention of other networks.

**Comment by NorbertZheng** [comment]
## AlexNet
### Architecture
![image](https://user-images.githubusercontent.com/37235023/236661541-dd573042-aaa5-4a40-89e8-700d25f6a8cd.png)
AlexNet.

AlexNet contains eight layers:

Input: 227×227×3 input images (224×224×3 sizes is mentioned in the paper and also in the figure, however, it is later pointed out that it should be 227, or 224×224×3 is padded during the 1st convolution.)

1st: Convolutional Layer: 2 groups of 48 kernels, size 11×11×3
(stride: 4, pad: 0)
Outputs 55×55 ×48 feature maps ×2 groups
Then 3×3 Overlapping Max Pooling (stride: 2)
Outputs 27×27 ×48 feature maps ×2 groups
Then Local Response Normalization
Outputs 27×27 ×48 feature maps ×2 groups

2nd: Convolutional Layer: 2 groups of 128 kernels of size 5×5×48
(stride: 1, pad: 2)
Outputs 27×27 ×128 feature maps ×2 groups
Then 3×3 Overlapping Max Pooling (stride: 2)
Outputs 13×13 ×128 feature maps ×2 groups
Then Local Response Normalization
Outputs 13×13 ×128 feature maps ×2 groups

3rd: Convolutional Layer: 2 groups of 192 kernels of size 3×3×256
(stride: 1, pad: 1)
Outputs 13×13 ×192 feature maps ×2 groups

4th: Convolutional Layer: 2 groups of 192 kernels of size 3×3×192
(stride: 1, pad: 1)
Outputs 13×13 ×192 feature maps ×2 groups

5th: Convolutional Layer: 256 kernels of size 3×3×192
(stride: 1, pad: 1)
Outputs 13×13 ×128 feature maps ×2 groups
Then 3×3 Overlapping Max Pooling (stride: 2)
Outputs 6×6 ×128 feature maps ×2 groups

6th: Fully Connected (Dense) Layer of
4096 neurons

7th: Fully Connected (Dense) Layer of
4096 neurons

8th: Fully Connected (Dense) Layer of
Outputs 1000 neurons (since there are 1000 classes)
Softmax is used for calculating the loss.

In total, there are 60 million parameters need to be trained !!!

**Comment by NorbertZheng** [comment]
### ReLU
![image](https://user-images.githubusercontent.com/37235023/236661821-6a757ce2-c64c-48b9-b8fb-367d96ed799a.png)

Before Alexnet, Tanh was used. ReLU is introduced in AlexNet.
 - ReLU is six times faster than Tanh to reach 25% training error rate.

**Comment by NorbertZheng** [comment]
### Multiple GPUs
At that moment, NVIDIA GTX 580 GPU is used which only got 3GB of memory. Thus, we can see in the architecture that they split into two paths and use 2 GPUs for convolutions. Inter-communications are only occurred at one specific convolutional layer.

Thus, using 2 GPUs, is due to memory problem, NOT for speeding up the training process.

With the whole network compared with a net with only half of kernels (only one path), Top-1 and top-5 error rates are reduced by 1.7% and 1.2% respectively.
 - More filters is better!!!

**Comment by NorbertZheng** [comment]
### Local Response Normalization
![image](https://user-images.githubusercontent.com/37235023/236662055-2b051270-9018-4201-ae03-2afde26d9763.png)
Normalization.

In AlexNet, local response normalization is used. It is different from the batch normalization as we can see in the equations. Normalization helps to speed up the convergence.
 - Nowadays, batch normalization is used instead of using local response normalization.

With local response normalization, Top-1 and top-5 error rates are reduced by 1.4% and 1.2% respectively.

**Comment by NorbertZheng** [comment]
### Overlapping Pooling
Overlapping Pooling is the pooling with stride smaller than the kernel size while Non-Overlapping Pooling is the pooling with stride equal to or larger than the kernel size.

With overlapping pooling, Top-1 and top-5 error rates are reduced by 0.4% and 0.3% respectively.

**Comment by NorbertZheng** [comment]
### Data Augmentation
Two forms of data augmentation.

**First: Image translation and horizontal reflection (mirroring)**
A random 224×224 is extracted from one 256×256 image plus horizontal reflection. The size of training set is increased by a factor of 2048. This can be calculated as follows:

By image translation: (256–224)²=32²=1024

By horizontal reflection: 1024 × 2 = 2048

At the test time, four corner patches plus the centre patch as well as their corresponding horizontal reflections (10 patches in total), are used for prediction, and get the average of all results to obtain the final classification result.

**Second: Altering the intensity**
PCA is perform on the training set. For each training image, add the quantity:

![image](https://user-images.githubusercontent.com/37235023/236662157-63befb6e-c227-4b82-8fa7-6fd977ebbb53.png)
Quantity of intensity altered.

where $p_{i}$ and $\lambda_{i}$ are $i$-th eigenvector and eigenvalue of the 3×3 covariance matrix of RGB pixel values, respectively, and αi is the random variable with mean 0 and standard variation 0.1.

By increasing the size of training set with data augmentation, Top-1 error rate is reduced by over 1%.

**Comment by NorbertZheng** [comment]
### Dropout
![image](https://user-images.githubusercontent.com/37235023/236662239-aaa9ec85-2376-4125-90ff-7130ee8341f9.png)
Dropout.

With the layer that using dropout, during training, each neuron has a probability not to contribute to feed forward pass and participate in backpropagation. Thus, each neuron can have a larger chance to be trained, and not to depend so much for some very “strong” neuron.

During test time, there will be no dropout.

In AlexNet, probability of 0.5 is used at the first two fully-connected layers. Dropout is a kind of regularization technique to reduce the overfitting.

**Comment by NorbertZheng** [comment]
### Other Details of Learning Parameters
Batch size: 128
Momentum v: 0.9
Weight Decay: 0.0005
Learning rate ϵ: 0.01, reduced by 10 manually when validation error rate stopped improving, and reduced by 3 times.

![image](https://user-images.githubusercontent.com/37235023/236662287-53c00988-31bf-44d6-99b4-529c67d4aeeb.png)
The update of momentum v and weight w.

Training set of 1.2 million images.
Network is trained for roughly 90 cycles.
Five to six days on two NVIDIA GTX 580 3GB GPUs.

**Comment by NorbertZheng** [comment]
### Results
![image](https://user-images.githubusercontent.com/37235023/236662311-5c4346bb-2ced-4b13-a068-87a41b11ba5b.png)
Error Rate in ILSVRC 2010.

 - For ILSVRC 2010, AlexNet got the Top-1 and top-5 error rates of 37.5% and 17.0% respectively, which outperforms other approaches.
 - Without ~~averaging 10 predictions over ten patches by data augmentation~~, AlexNet only got the Top-1 and top-5 error rates of 39.0% and 18.3% respectively.

![image](https://user-images.githubusercontent.com/37235023/236662443-52be9a4e-e926-48d6-a326-13d3bb854a3c.png)
Error Rate in ILSVRC 2012.

 - By 1 AlexNet (1 CNN), the validation error rate is 18.2%.
 - By Averaging the prediction from 5 AlexNet (5 CNNs), the error rate is reduced to 16.4%. This is a kind of boosting technique already used in LeNet for digit classification.
 - By adding one more convolutional layer to AlexNet (1 CNN*), the validation error rate is reduced to 16.6%.
 - By Averaging the prediction from 2 modfiied AlexNet and 5 original AlexNet (7 CNNs*), the validation error rate is reduced to 15.4%.

![image](https://user-images.githubusercontent.com/37235023/236662544-6ffde65e-51f4-4db1-9033-b28c727622f3.png)
Some Top-5 results by AlexNet.

**Comment by NorbertZheng** [comment]
Boosting!!! Boosting!!!

**Comment by NorbertZheng** [comment]
## CaffeNet
CaffeNet is a 1-GPU version of AlexNet. The architecture is:
![image](https://user-images.githubusercontent.com/37235023/236662605-10f00b63-9ec4-483a-9c18-941f325ea668.png)
CaffeNet.

We can see that the 2 paths in AlexNet are combined to become one path.

It is noted that for early version of CaffeNet, the order of pooling and normalization layers is reversed, this is by accident. But in the current version of CaffeNet provided by Caffe, it has already provided the Caffenet with the correct order of pooling and normalization layers.

**Comment by NorbertZheng** [comment]
By investigating each component one by one, we can know the effectiveness of each component. : )

If interested, there is also a tutorial about [CaffeNet quick setup using Nvidia-Docker and Caffe](https://medium.com/@sh.tsang/very-quick-setup-of-caffenet-alexnet-for-image-classification-using-nvidia-docker-2-0-c3b75bb8c7a8) [3].

**Comment by NorbertZheng** [comment]
## References
 - [2012 NIPS] [AlexNet] [ImageNet Classification with Deep Convolutional Neural Networks](https://papers.nips.cc/paper/4824-imagenet-classification-with-deep-convolutional-neural-networks.pdf).
 - [2014 ACM MM] [CaffeNet] [Caffe: Convolutional Architecture for Fast Feature Embedding](https://ucb-icsi-vision-group.github.io/caffe-paper/caffe.pdf).
 - [VERY QUICK SETUP of CaffeNet (AlexNet) for Image Classification Using Nvidia-Docker 2.0 + CUDA + CuDNN + Jupyter Notebook + Caffe](https://medium.com/@sh.tsang/very-quick-setup-of-caffenet-alexnet-for-image-classification-using-nvidia-docker-2-0-c3b75bb8c7a8).
 - ILSVRC [ImageNet Large Scale Visual Recognition Competition](http://www.image-net.org/challenges/LSVRC/).

Also, Read
 - [Copy Trading](https://medium.com/coinmonks/top-10-crypto-copy-trading-platforms-for-beginners-d0c37c7d698c) | [Crypto Tax Software](https://medium.com/coinmonks/crypto-tax-software-ed4b4810e338)
 - [Grid Trading](https://coincodecap.com/grid-trading) | [Crypto Hardware Wallet](https://medium.com/coinmonks/the-best-cryptocurrency-hardware-wallets-of-2020-e28b1c124069)
 - [Crypto Telegram Signals](http://top%204%20telegram%20channels%20for%20crypto%20traders/) | [Crypto Trading Bot](https://medium.com/coinmonks/crypto-trading-bot-c2ffce8acb2a)
 - [Best Crypto Exchange](https://medium.com/coinmonks/crypto-exchange-dd2f9d6f3769) | [Best Crypto Exchange in India](https://medium.com/coinmonks/bitcoin-exchange-in-india-7f1fe79715c9)
- [Best Crypto APIs](https://medium.com/coinmonks/best-crypto-apis-for-developers-5efe3a597a9f) for Developers
 - Best [Crypto Lending Platform](https://medium.com/coinmonks/top-5-crypto-lending-platforms-in-2020-that-you-need-to-know-a1b675cec3fa)
 - An ultimate guide to [Leveraged Token](https://medium.com/coinmonks/leveraged-token-3f5257808b22)
